File: facilitator-solutions/javascript/rung-5-bitset.js
// Rung 5 - bit packing. Includes rungs 1 to 4.
//
// One bit per odd candidate. Eight times smaller again, sixteen times smaller than
// the Uint8Array, and a sixty-fourth of where the JavaScript starting point began.
//
// JAVASCRIPT'S BITWISE OPERATORS ARE 32 BIT, so this wants a Uint32Array with
// >>> 5 and & 31, not the 64-bit word layout the .NET versions use. BigUint64Array
// drags in BigInt and is slower, not faster.
//
//   word index    i >>> 5
//   bit within    i & 31
//
// JavaScript masks the shift count for you, so `1 << i` already behaves as
// `1 << (i & 31)`. Go does not, which is why the Go version writes the mask out.
//
// A set bit still means prime here, so the words are filled with ones and the tail
// word's spare bits are set too. results is bounded by count, so they are never
// read - rung 6 is where that becomes a bug.
//
// Expect this to add less here than it does in C#. A Uint8Array is already fast
// and already small enough at a million.

export const getSieve = (sieveSize) => {
	const count = (sieveSize + 1) >> 1;
	const words = new Uint32Array((count + 31) >>> 5);
	words.fill(0xffffffff);

	const limit = Math.floor(Math.sqrt(sieveSize));

	return {
		run: () => {
			for (let factor = 3; factor <= limit; factor += 2) {
				const f = factor >> 1;

				if ((words[f >>> 5] & (1 << f)) !== 0) {
					for (let q = (factor * factor) >> 1; q < count; q += factor) {
						words[q >>> 5] &= ~(1 << q);
					}
				}
			}
		},
		results: () => {
			const primes = [];
			if (sieveSize >= 2) {
				primes.push(2);
			}
			// !== 0 and not > 0: a bit test with bit 31 set comes back negative.
			for (let i = 1; i < count; i++) {
				if ((words[i >>> 5] & (1 << i)) !== 0) {
					primes.push(2 * i + 1);
				}
			}
			return primes;
		},
	};
};

File: facilitator-solutions/go/rung-5-bitset.go
package main

import "math"

// Rung 5 - bit packing. Includes rungs 2 to 4.
//
// One bit per odd candidate, packed into 64-bit words. Eight times smaller again,
// sixteen times smaller than the original byte-per-number slice.
//
//	word index    i >> 6
//	bit within    i & 63
//
// GO DOES NOT MASK SHIFT COUNTS. In C# and JavaScript, 1 << i quietly behaves as
// 1 << (i & wordBits-1). In Go a shift count at or beyond the word width produces
// zero, silently, so the mask has to be written out: 1 << (uint(q) & 63). A Go
// bitset that "nearly works" is usually this. Do not tidy it away.
//
// A set bit still means prime here, so the words are filled with ones. The tail
// word's spare bits are set too, and result is bounded by count so they are never
// read - rung 6 is where that turns into a bug.

type primeSieve struct {
	words []uint64
	size  int
}

func getSieve(sieveSize int) primeSieve {
	count := (sieveSize + 1) / 2
	words := make([]uint64, (count+63)/64)
	for i := range words {
		words[i] = ^uint64(0)
	}
	return primeSieve{words: words, size: sieveSize}
}

func (sieve primeSieve) run() {
	limit := int(math.Sqrt(float64(sieve.size)))
	count := (sieve.size + 1) / 2

	for factor := 3; factor <= limit; factor += 2 {
		f := factor >> 1

		if sieve.words[f>>6]&(1<<(uint(f)&63)) != 0 {
			for q := (factor * factor) >> 1; q < count; q += factor {
				sieve.words[q>>6] &^= 1 << (uint(q) & 63)
			}
		}
	}
}

func (sieve primeSieve) result() []int {
	var primes []int
	count := (sieve.size + 1) / 2

	if sieve.size >= 2 {
		primes = append(primes, 2)
	}

	for i := 1; i < count; i++ {
		if sieve.words[i>>6]&(1<<(uint(i)&63)) != 0 {
			primes = append(primes, 2*i+1)
		}
	}

	return primes
}
